### ISSUE TITLE
Tracking Issue for atomic_from_mut

### ISSUE THREAD

**Issue by KodrAus** [rust-lang/rust#76314]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/issues/76314)*

Feature gate: `#![feature(atomic_from_mut)]`

### Public API

```rust
impl AtomicU32 {
    pub fn from_mut(v: &mut u32) -> &mut Self;
    pub fn from_mut_slice(v: &mut [u32]) -> &mut [Self];
    pub fn get_mut_slice(this: &mut [Self]) -> &mut [u32];
}

// And same for AtomicBool, AtomicU8, AtomicU16, AtomicU64, AtomicI8, AtomicI16, AtomicI32, AtomicI64, and AtomicPtr
```

Each one is gated on `#[cfg(target_has_atomic_equal_alignment = "..")]` and is only available on platforms where `Atomic<size>` has the same alignment as `u<size>`.

### Steps / History

- [x] Implementation: #74532
- [x] #92671
- [x] #94384
- [x] https://github.com/rust-lang/rust/pull/94816
- [ ] Final commenting period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- ~Should these return `&mut Self` instead, such that it's the exact inverse of `get_mut`?~ &mdash; **yes:** #92671
- ~Also add `from_mut_slice`?~ &mdash; **yes:** #94384

**Comment by RalfJung** [comment]
For some types, these functions introduce a large risk of accidentally writing non-portable code: when I am on x86_64, I might not realize that by calling `AtomicU64::from_mut` I am excluding platforms like x64_32 where `u64` has alignment less than 8.

**Comment by m-ou-se** [comment]
But the same holds for the other atomics operations, right? Using 64-bit atomic cas operations will also exclude a lot of platforms.

The alternative is that people write this manually using transmute. That code will be equally non-portable, except it'll compile just fine and silently cause UB.

That said, it might be good to extend the note in the documentation to mention platforms like x86 and x64_32 specifically.

**Comment by RalfJung** [comment]
Yeah, I am not saying this is a blocker. This could be something that good documentation can solve.

**Comment by RustyYato** [comment]
Can `Atomic*::from_mut`'s signature be changed to `fn(&mut _) -> &mut Self` instead of `fn(&mut _) -> &Self`. This way `from_mut` and `get_mut` are inverses. (I view `Cell::from_mut`'s signature as a wart, that can't be fixed, but `Atomic*::from_mut` can be changed).

**Comment by m-ou-se** [comment]
@RustyYato Is there a use case for that? There's not much more you can do with a `&mut Atomic` other than `get_mut()`.

**Comment by RustyYato** [comment]
@m-ou-se No specific use-case, it just seems odd to me that `from_mut` and `get_mut` are not inverses.

**Comment by m-ou-se** [comment]
Okay. I've added your suggestion to the unresolved questions to resolve before stabilizing this.

**Comment by the8472** [comment]
> Should these return &mut Self instead, such that it's the exact inverse of get_mut?

I think the following two make sense.

* `from_mut(&mut _) -> &mut Self`, inverse of `get_mut`
* `unsafe from_mut_ptr(*mut _) -> &Self`, inverse of `as_mut_ptr` and does not allow `get_mut`

**Comment by xelahi** [comment]
> `unsafe from_mut_ptr(*mut _) -> &Self`, inverse of `as_mut_ptr` and does not allow `get_mut`

What would be the safety requirements on `from_mut_ptr`? Would overlapping atomics of different sizes be alllowed?

Overlapping atomics of different sizes is probably an obscure use case, but it would be really helpful in some situations. However, Rust's memory model (i.e., C++ memory model) does not appear to account for that possibility.

**Comment by RalfJung** [comment]
Such averlapping atomics are called "mixed accesses" in the weak memory literature, and indeed most memory models do not support them.

**Comment by Vurich** [comment]
What is the reasoning behind it being invalid to cast from `&usize` to `&AtomicUsize` or the equivalents for other types? I would assume the UB invariants are the same as for `&T` to `&UnsafeCell<T>` (or, equivalently, `&T` to `&Cell<T>`) but I can't find any info on why that is invalid either.

**Comment by CryZe** [comment]
> What is the reasoning behind it being invalid to cast from `&usize` to `&AtomicUsize` or the equivalents for other types?

&usize is immutable, the compiler is allowed to assume the value behind the reference will never change. &AtomicUsize is "mutable" through it's atomic operations. So any sort of cast like that is unsound.

**Comment by Vurich** [comment]
Ah right, that makes sense. It's not the same as a `&T -> &mut T` cast but it causes the same issues with code motion.

**Comment by WaffleLapkin** [comment]
These methods were in `std` for quite a while now, can we move towards stabilization? 

As a first step, I can make a PR that changes the output from `&Self` to `&mut Self`. It doesn't seem there are any concerns with that and this makes the API nicer (by making `from_mut` and `get_mut` inverses of each other).

**Comment by RalfJung** [comment]
One concern we might have is that AFAIK C++ does not have a comparable API, so this could move us into a situation where we cannot just copy the C++ memory model for atomics.

**Comment by WaffleLapkin** [comment]
I'm not very familiar with C++, but after a quick scan of its [atomic docs](https://en.cppreference.com/w/cpp/atomic/atomic) it seems like C++ doesn't have `get_mut` either, so we are already in the said situation, right?

**Comment by RalfJung** [comment]
C++ has `atomic_ref`.

But I think I just realized that `atomic_ref` is like `from_mut`, so the problematic one is the one we already stabilized...

**Comment by comex** [comment]
However, `atomic_ref` is not guaranteed to have a trivial destructor or copy constructor, and all accesses to the object while the `atomic_ref` is alive are required to be through `atomic_ref` itself.  Therefore, it almost seems like `atomic_ref` could be implemented "pathologically" as:
- the constructor takes a copy of the object and store it in the `atomic_ref` itself;
- all atomic methods refer to the copy rather than the original pointer;
- the destructor puts the object back.

In reality, this isn't quite allowed, because you can separately make multiple `atomic_ref`s to the same pointer, and them access all of them concurrently.  All accesses are required to be through `atomic_ref`, but not through the *same* `atomic_ref`.

However, this can be fixed by adding a layer of indirection.  Instead of `atomic_ref` storing a copy of the object directly, it would be essentially `shared_ptr<atomic_ref_backing>`, where `atomic_ref_backing` stores a copy of the object, and there's a global hash map from pointers to `atomic_ref_backing`s so that creating multiple `atomic_ref`s from the same pointer gives you references to the same `atomic_ref_backing`.

I'm not sure whether the fact this is possible is an intentional choice to support architectures that can't manipulate arbitrary memory atomically, or just a coincidence.  Either way, it means that `AtomicFoo::from_mut` is a bit more demanding of the platform than `atomic_ref`, since it returns a reference, which is guaranteed to be trivially copyable and droppable, ruling out such an implementation.

Then again, `AtomicFoo` in general is already more demanding than C++ `atomic`, as it is required to be lock-free while C++ `atomic` is not.

**Comment by bonzini** [comment]
Is it possible to write a signature that *consumes* the &mut? The point of from_mut is that an AtomicUxxx *can* be accessed safely from multiple threads, without causing UB. It supports interior mutability, so it should be possible to get a shared reference to an atomic without using unsafe. But that's only possible if one cannot write

```rust
    let x: usize = 42;
    let ref = &x;
    let aref = AtomicUsize::from_mut(ref); // Shared ref
    // Send aref around...
    *ref = 53;    // Undefined behavior!!!
```

If that's not possible then yeah it has to be unsafe.

**Comment by WaffleLapkin** [comment]
@bonzini borrow checker will stop you from doing any undefined behavior here :)

`AtomicUsize::from_mut` returns a reference, so the lifetime is preserved, `xref` will be frozen while `xaref` is alive.
You can use [playground] to see that this code is rejected:
```rust
let mut x: usize = 42;
let xref = &mut x;
let xaref = AtomicUsize::from_mut(xref);

// Send xaref around...

*xref = 53;

dbg!(xaref);
```
```text
error[[E0506]](https://doc.rust-lang.org/nightly/error-index.html#E0506): cannot assign to `*xref` because it is borrowed
  --> src/main.rs:10:5
   |
6  |     let xaref = AtomicUsize::from_mut(xref);
   |                                       ---- borrow of `*xref` occurs here
...
10 |     *xref = 53;
   |     ^^^^^^^^^^ assignment to borrowed `*xref` occurs here
11 |     
12 |     dbg!(xaref);
   |          ----- borrow later used here
```

[playground]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=395f854510269939f1327f3ecbf41f69

**Comment by RalfJung** [comment]
> Is it possible to write a signature that consumes the &mut?

Your example uses an &, not an &mut, so I am a bit confused by the question in general.

**Comment by bonzini** [comment]
> Your example uses an &, not an &mut

Don't write Rust on a phone is the logical conclusion of that (I edited now). Anyway, then I think from_mut should return a shared reference.

**Comment by RalfJung** [comment]
But why? There is no benefit from that, only less (well-defined) code that can compile.

**Comment by bonzini** [comment]
> There is no benefit from that

The benefit is being able to go safely from single-thread mutable `&mut u32` to multi-thread mutable `&AtomicU32`. For example you can get a `&mut u32`, let multiple threads do their things on it and then give it back to the caller. (Of course the benefit is more visible with large slices than with a single value).

[Here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=78e178265003e1f0e45c31017819bba2) is a playground link that demonstrate this possibility.

**Comment by RalfJung** [comment]
> The benefit is being able to go safely from single-thread mutable &mut u32 to multi-thread mutable &AtomicU32.

You can do that with the current signature. `&mut AtomicU32` coerces to `&AtomicU32`.

So, what you want can already be done. No need to make `from_mut` return a shared reference.

**Comment by bonzini** [comment]
Ah, of course (`let result_atomic: &AtomicU32 = AtomicU32::from_mut(result);`). Sorry for missing the obvious.

**Comment by zseri** [comment]
could the alignment requirement be integrated into the type signatures for the cases where that is problematic? (perhaps via some `Option` checked-construction function)

**Comment by WaffleLapkin** [comment]
I don't think that this function makes sense as a fallible one. I don't see a usecase where you'd be able to recover somehow/having a function work only sometimes is valuable...

**Comment by zseri** [comment]
this is why I'd like to see it integrated into the type signatures, to make sure it can't even be called with under-aligned objects.

**Comment by WaffleLapkin** [comment]
The function is unavailable on targets where `align_of::<T>() != align_of::<AtomicT>()`, so it is impossible to call with under-aligned objects.

**Comment by zseri** [comment]
It would be nice to have an alternative on platforms where the alignment doesn't match *per default*, that is, a way to force the alignment and also specify that in the type signature would be nice (on platforms where the alignment wouldn't be widened, that would be a no-op)

**Comment by WaffleLapkin** [comment]
We could have some sort of `struct AlignedTo<T, const ALIGN: usize>(...);`, but that's not quite simple (you need to set the alignment somehow and also probably use it as `AlignedTo<u32, align_of::<AtomicU32>()>`), I don't think the we are ready for this (yet). So, I'd say it would be nice to have eventually, but it's a task for the future....

**Comment by uablrek** [comment]
Have you considered shared memory (Linux)?

I am writing a traffic generator (in rust) where the statistics is stored in shared memory. An independent presentation program (in any language) can map the shared memory and present statistics in real time.

The traffic generator is multi-threaded (tokio), so counters must be updated atomically. Example:
```rust
// The "stats" struct is mapped to shared memory
let bytes_sent = AtomicU64::from_mut(&mut stats.bytes_sent);
bytes_sent.fetch_add(bytes, Ordering::Relaxed);
```
This works fine, but I must use `rustup default nightly` (no big deal since I am experimenting).

Is there a way to accomplish this in a way that can be used in "stable"?
(perhaps some trick with Box<> and from_ptr?)

Or shall I wait until this feature makes it to "stable".

**Comment by RalfJung** [comment]
If you are using shared memory, and you ever have a mutable reference to that memory, something already went terribly wrong. Mutable references must be unique, which implies "not shared".

**Comment by uablrek** [comment]
Um, are you saying that the use of Linux shared memory is forbidden in rust? Or at least writing to it?

**Comment by the8472** [comment]
You shouldn't use `&mut` for those and use raw pointers and `AtomicXXX::from_ptr` instead.

**Comment by RalfJung** [comment]
Correct, shared memory has to be behind a raw pointer or behind a shared reference inside an UnsafeCell (or immutable, then it can be a regular shared reference). It's the same rules as when sharing memory between multiple threads inside a single process, so this should not be surprising.

**Comment by uablrek** [comment]
Changed to:
```rust
let bytes_sent = unsafe {
    AtomicU64::from_ptr(&mut stats.bytes_sent as *mut u64)
};
```
and switched back to `rustup default stable`. Works fine. Thanks a lot @the8472

**Comment by RalfJung** [comment]
> `&mut stats.bytes_sent`

You're still creating a mutable reference here, so this code still breaks the aliasing rules.
You can use `addr_of_mut!(stats.bytes_sent)` to avoid that.

Also, if `stats` is a mutable reference, then that should be changed as well.
But anyway that's all off-topic in this issue; feel free to ask further questions on [Zulip](https://rust-lang.zulipchat.com/). :)

**Comment by uablrek** [comment]
> But anyway that's all off-topic in this issue; feel free to ask further questions on [Zulip](https://rust-lang.zulipchat.com/). :)

Will do. You are right, I thought it wasn't possible to use an existing u64 for atomic operations without "nightly". Hence my comment in the issue. Sorry.

Updated again to;
```rust
let bytes_sent = unsafe {
    AtomicU64::from_ptr(addr_of_mut!(stats.bytes_sent))
};
```
Thanks.

**Comment by joshtriplett** [comment]
Starting an FCP to stabilize:

@rfcbot merge

**Comment by rust-rfcbot** [comment]
@joshtriplett has proposed to merge this. The next step is review by the rest of the tagged team members:

* [x] @Amanieu
* [x] @BurntSushi
* [x] @dtolnay
* [x] @joshtriplett
* [ ] @the8472

Concerns:

* ~~target_has_atomic_equal_alignment~~ resolved by https://github.com/rust-lang/rust/issues/76314#issuecomment-4724905850
* ~~use case~~ resolved by https://github.com/rust-lang/rust/issues/76314#issuecomment-3493984618

Once a majority of reviewers approve (and at most 2 approvals are outstanding), this will enter its final comment period. If you spot a major issue that hasn't been raised at any point in this process, please speak up!

See [this document](https://github.com/rust-lang/rfcbot-rs/blob/master/README.md) for info about what commands tagged team members can give me.

**Comment by Amanieu** [comment]
One issue that came up in the meeting is whether anyone actually had a use case for these methods. Could someone please explain how they intend to use these? I am uncomfortable stabilizing these if nobody is going to use them.

@rfcbot concern use case

**Comment by programmerjake** [comment]
one use case from the equivalent C++ feature: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0019r8.html#atomic-operations-on-members-of-a-very-large-array

**Comment by BurntSushi** [comment]
@rfcbot concern target_has_atomic_equal_alignment

**Comment by Mark-Simulacrum** [comment]
@BurntSushi I think you can resolve the concern; the cfg is now stable (stabilization PR - https://github.com/rust-lang/rust/pull/155006 has merged).

**Comment by BurntSushi** [comment]
Thanks!

@rfcbot resolve target_has_atomic_equal_alignment

**Comment by rust-rfcbot** [comment]
:bell: **This is now entering its final comment period**, as per the [review above](https://github.com/rust-lang/rust/issues/76314#issuecomment-3486862895). :bell:

**Comment by rust-rfcbot** [comment]
The final comment period, with a disposition to **merge**, as per the [review above](https://github.com/rust-lang/rust/issues/76314#issuecomment-3486862895), is now **complete**.

As the automated representative of the governance process, I would like to thank the author for their work and everyone else who contributed.
